accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -350,7 +350,11 @@ initializer expression and is not marked `late` is used before it is definitely
 assigned (see Definite Assignment below).
 
 It is an error if the body of a method, function, getter, or function expression
-with a potentially non-nullable return type **may complete normally**.
+with a potentially non-nullable return type **may complete normally** or may
+return without an object.
+
+It is an error if the body of a method, function, getter, or function expression
+with return type `Never` may return with an object.
```

Comment:
I'm not sure that this is necessary, nor desirable.

```dart
Never bar() => throw "Unreachable";
Never foo() {return bar();}
```

is a good program, and should be supported, I think.

```dart
dynamic bar() => throw "Unreachable";
Never foo() {return bar();}
```

is a bad program, but the reason has nothing to do with `Never`.  There's just an implicit downcast there from `dyanamic` to `Never`.

I think this program:
```dart
Never foo() {return;}
```

does need some further specification.

Reply:
```dart
Never bar() => throw "Unreachable";
Never foo() { return bar(); }
```

This can be defended because it should always be OK to return something of type `T` when the return type is `T`. It does have some of the same weirdness as `void f() { return print("Hello!"); }`, in that it is misleading to read `return ..` and think "a value is returned here". But we allow that exactly because it returns `void` to `void`, so we should probably accept a similar weirdness with `Never` as well.

```dart
dynamic bar() => throw "Unreachable";
Never foo() { return bar(); }
```

However, I'm not convinced that this is a bad program. We just decided that it will be allowed with null-safety enabled to implicitly cast from `dynamic` to any type. So why should a cast from `dynamic` to `Never` be an exception?

I argue in #913 that we would then trust the expression of static type `dynamic` to loop forever, or to throw. But this is no more risky than trusting an expression of static type `dynamic` to evaluate to an object whose interface has any given member, say, with `d.fooBar(42)`.

In any case, if that expression of static type `dynamic` does _not_ live up to the expectation (and it does evaluate to an object), then there will be a dynamic check, and we will get an exception at run time.

```dart
Never foo() { return; }
```

As I argued above (at line 354) that this is already covered by the rules in `\section{Return}` of the language specification, so I don't think we need more to make that an error.